File: scratch/extracted/dijkstras/optimize.py
def networkDelayTime(times: list[list[int]], n: int, k: int) -> int:
    # Create an adjacency list (dictionary) to represent the graph.
    # Key: source node, Value: list of (neighbor, weight) pairs.
    edges = {}
    for u, v, w in times:
        if u not in edges:
            edges[u] = []
        edges[u].append((v, w))

    # Initialize the min-heap with the starting node and its initial time (0).
    # The min-heap stores (time, node) pairs, where time is the time to reach the node from the source.
    minHeap = [(0, k)]  # (time, node)

    # 'visit' is a set to keep track of visited nodes to avoid cycles and redundant processing.
    visit = set()

    # 't' will store the maximum time it takes to reach all reachable nodes from the source.
    t = 0

    # Dijkstra's algorithm: iterate while the min-heap is not empty.
    while minHeap:
        # Sort the minHeap to get the smallest element. Note: heapq would be more efficient here for a true heap implementation.
        minHeap.sort()

        # Get the node with the smallest time from the min-heap.
        w1, n1 = minHeap.pop(0)

        # If the node has already been visited, skip it.
        if n1 in visit:
            continue

        # Mark the node as visited.
        visit.add(n1)

        # Update the maximum time 't' seen so far.  This represents the longest shortest path.
        t = max(t, w1)

        # Iterate through the neighbors of the current node.
        if n1 in edges:
            for n2, w2 in edges[n1]:
                # If a neighbor has not been visited, add it to the min-heap with its updated time.
                if n2 not in visit:
                    minHeap.append((w1 + w2, n2))

    # After the loop finishes, check if all nodes have been visited.
    # If yes, return the overall time 't'. Otherwise, return -1, indicating that not all nodes are reachable.
    return t if len(visit) == n else -1

File: scratch/extracted/dijkstras/optimize.cpp


class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {

        // Build adjacency list representing the graph.
        // The keys are the source nodes, and the values are vectors of pairs.
        // Each pair represents a destination node and the weight of the edge.
        unordered_map<int, vector<pair<int,int>>> edges;

        // Iterate through the 'times' vector to populate the adjacency list.
        for (auto &time : times) {
            int u = time[0]; // Source node.
            int v = time[1]; // Destination node.
            int w = time[2]; // Weight of the edge from u to v.
            edges[u].push_back({v, w}); // Add the edge to the adjacency list.
        }

        // Min heap (priority queue) to store nodes to visit, prioritized by distance.
        // The pair stores {distance, node}.
        // 'greater<pair<int,int>>' makes it a min-heap.
        priority_queue<
            pair<int,int>,
            vector<pair<int,int>>,
            greater<pair<int,int>>
        > minHeap;

        // Add the starting node 'k' to the min-heap with a distance of 0.
        minHeap.push({0, k});

        // 'visit' set to keep track of visited nodes.
        set<int> visit;

        // 't' to store the maximum time taken to reach all nodes.
        int t = 0;

        // Dijkstra's algorithm.
        while (!minHeap.empty()) {

            // Get the node with the smallest distance from the min-heap.
            auto [w1, n1] = minHeap.top(); // w1: distance, n1: node.
            minHeap.pop();

            // If the node has already been visited, skip it.
            if (visit.count(n1)) continue;

            // Mark the current node as visited.
            visit.insert(n1);

            // Update the maximum time taken so far.
            t = max(t, w1);

            // Iterate through the neighbors of the current node.
            for (auto &[n2, w2] : edges[n1]) { // n2: neighbor, w2: weight of edge from n1 to n2.
                // If the neighbor has not been visited, add it to the min-heap.
                if (!visit.count(n2)) {
                    minHeap.push({w1 + w2, n2}); // Distance to neighbor is the distance to current node + edge weight.
                }
            }
        }

        // If all nodes have been visited, return the maximum time taken. Otherwise, return -1.
        return visit.size() == n ? t : -1;
    }
};